Turn into ES module

nikku opened the pull request:
### What is inside

Turns this library into a true ES module - a breaking change.

* Preserves module imports, the most standard way of using diagram-js
  
  ```
  import SelectionModule from 'diagram-js/lib/features/selection';
  ```

* Requires files to be imported with extension, i.e. 

  ```
  import { getParents } from 'diagram-js/lib/util/Elements.js'
  ```

---

To aid the migration of libraries using diagram-js, consider our `esm` [codemod](https://github.com/bpmn-io/codemods) - it migrates your library to consume utilities via file extensions:

```
npx @bpmn-io/codemods esm --dry-run your-library
```

---

Cf. https://github.com/bpmn-io/diagram-js/pull/870#issuecomment-1993684325 for full discussion of options.

---

Depends on https://github.com/bpmn-io/diagram-js/pull/869

Closes #863

nikku commented:
This PR poses the question: How do we want downstream libraries consume diagram-js?

In bpmn-js and other libraries we currently consume like this:

```
import ModelUtil from 'diagram-js/lib/ModelUtil';
import zoomScroll from 'diagram-js/lib/navigation/zoomscroll';
```

Also possible right now, explicit import:

```
import ModelUtil from 'diagram-js/lib/ModelUtil.js';
import zoomScroll from 'diagram-js/lib/navigation/zoomscroll/index.js';
```

As we create ES modules with an `exports` declaration we have to be _explicit_ about what we export (and hence can be imported). 

Proposed in this PR we force our users to switch to explicit imports, which is a major breaking change for all consumers (cf. bpmn-js test :arrow_down:):

```
ERROR in ./test/util/custom-rules/CustomRules.js 3:0-70
Module not found: Error: Package path ./lib/features/rules/RuleProvider is not exported from package [PRJ]/bpmn-js/node_modules/diagram-js (see exports field in [PRJ]/bpmn-js/node_modules/diagram-js/package.json)
 @ ./test/util/custom-rules/index.js 1:0-40 5:25-36
 @ ./test/spec/features/context-pad/ContextPadProviderSpec.js 27:0-59 39:4-21
 @ ./test/ sync (spec%7Cintegration).*Spec\.js$ ./spec/features/context-pad/ContextPadProviderSpec.js test/spec/features/context-pad/ContextPadProviderSpec.js
 @ ./test/testBundle.js 1:15-74

webpack 5.88.1 compiled with 273 errors and 39 warnings in 1670 ms
```

So we have three options:

* Build `exports` in a way that current import scenario works
* Switch to semantic ESM (import via file extension)
* Do not define `exports` field

barmac commented:
If we are to ship this, we could consider writing a codemod to make the migration easier (for example with https://github.com/facebook/jscodeshift).

nikku committed:
chore: turn karma configuration into CJS

nikku committed:
chore!: turn project into ESM module

nikku committed:
test: adjust bundles to use ESM syntax

nikku committed:
test: use package references for cross-module referencing

This ensures tooling can natively understand the structure - following
simple ESM resolution rules.

nikku committed:
test: remove custom resolution hack

nikku committed:
fix: correct type reference

nikku commented:
Adjusted this PR on top of https://github.com/bpmn-io/diagram-js/pull/869 - with the following changes:

* We export modules - this is the standard way of consumers currently using our libraries
* Still a breaking change - utilities have to be imported with `.js` extensions

barmac commented on test/util/MockEvents.js:
```diff
@@ -2,18 +2,18 @@ import { assign } from 'min-dash';
  * specified canvas local coordinates.
  *
- * @param {Point} point of the event local the canvas (closure)
+ * @param {Point} position of the event local the canvas (closure)
```
```suggestion
 * @param {Point} position of the event local to the canvas (closure)
```

nikku committed:
docs: improve parameter documentation

Co-authored-by: Maciej Barelkowski <28307541+barmac@users.noreply.github.com>

nikku commented on eslint.config.mjs:
```diff
@@ -48,6 +49,17 @@ export default [
+    rules: {
+      'import-x/extensions': 'off'
+    }
+  },
```
Wrong - it includes `lib` in the process.

nikku committed:
16.0.0-esm.0

nikku committed:
chore: rename eslint file

nikku committed:
chore: explicitly expose modules

* allows consumers to import as they are used to
* adjust tests (JS, TS) to use these exports - test the contract

nikku committed:
16.0.0-esm.1

nikku committed:
16.0.0-esm.2
